Handle Placement When Multiple Adjacent Factories

ron-murhammer opened the pull request:
Addresses #350

**Changes**
Prompt user for which producer territory (factory) to use when producing units in sea zones that have multiple adjacent factories. This is particularly important when users can move existing air units to carriers as they only can be moved from the producer territory. Though it is generally useful as otherwise players need to make sure to set land units first so sea units don't utilize the production from the factory.

**Testing**
Open any map that has naval units and make sure there is a sea territory next to at least 2 of your factories. Purchase at least 1 sea unit and place it into that territory.
Here is a save game where you can place a carrier in sea zone 5 (baltic sea): 
[test.zip](https://github.com/triplea-game/triplea/files/1034872/test.zip)


**New Popup**
![capture](https://cloud.githubusercontent.com/assets/1427689/26537304/d3e8888a-4403-11e7-9e0b-b1f4e691ca88.PNG)

ron-murhammer commented on src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:
```diff
@@ -204,23 +204,38 @@ public String placeUnits(final Collection<Unit> units, final Territory at, final
+          break;
+        }
+      }
+      producers.remove(producer);
```
The 15 lines above this comment are the core of the change. It checks if there are multiple factories and if so then call the player to choose one. It loops until all of the units are placed or all territories are utilized.

ron-murhammer commented on src/main/java/games/strategy/triplea/ui/TripleAFrame.java:
```diff
@@ -1979,6 +1979,44 @@ public void actionPerformed(final ActionEvent e) {
+   *        - territory units are being placed
+   * @return territory selected to use for production
+   */
+  public Territory selectProducerTerritoryForUnits(final Collection<Territory> candidates,
```
Here is the method that actually puts of the popup to the user with a list of territories to select from.

RoiEXLab commented:
@ron-murhammer Would you mind rebasing onto the latest master?
Not only are merge conflicts showing up, but it also makes the diff really hard to read since we can't see what changes were already made in #1763

RoiEXLab commented:
@ron-murhammer 
Also if you have the time, would you mind looking at some of the PRs at the website repo?
Most of the changes in the pending PRs are pretty basic: https://github.com/triplea-game/triplea-game.github.io/pulls

RoiEXLab commented on src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:
```diff
@@ -204,23 +204,38 @@ public String placeUnits(final Collection<Unit> units, final Territory at, final
+    unitsLeftToPlace.sort(getUnitConstructionComparator());
+
+    while (true) {
+      if (unitsLeftToPlace.isEmpty() || producers.isEmpty()) {
```
This if statement should be merged with the while loop

ron-murhammer replied:
Done.

RoiEXLab commented on src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:
```diff
@@ -204,23 +204,38 @@ public String placeUnits(final Collection<Unit> units, final Territory at, final
+        continue;
+      } else if (maxPlaceable == 0) {
+        maxPlaceable = 1;
+      }
```
I'd use 2 nested if statements here:
```
if(maxPlaceable == 0){
  if(bidMode == BidMode.NOT_BID){
    continue;
  }
  maxPlaceable = 1;
}
```

ron-murhammer replied:
Done.

ron-murhammer commented:
@RoiEXLab Yeah, didn't realize til I pushed it up here that it needed rebased so just took care of that. Addressed your code comments as well. I'll leave this PR up here for a day or so to see if anyone else has any feedback.

ssoloff commented on src/main/java/games/strategy/triplea/ui/TripleAFrame.java:
```diff
@@ -1979,6 +1979,44 @@ public void actionPerformed(final ActionEvent e) {
+      final JScrollPane scroll = new JScrollPane(list);
+      panel.add(scroll, BorderLayout.CENTER);
+      panel.add(BorderLayout.NORTH, new JLabel("Producer territory for units in " + unitTerritory.getName()));
+      final String[] options = {"OK", "Canel"};
```
`"Canel"` --> `"Cancel"`

Note, however, that passing `null` for the `options` parameter will use the appropriate locale-specific button labels.  The Javadocs for `JOptionPane#showOptionDialog()` seem to imply this is only done if `optionType` is `YES_NO_OPTION` or `YES_NO_CANCEL_OPTION`, but it appears we've used `null` in other locations where `OK_CANCEL_OPTION` is specified (e.g. `SetPasswordAction#actionPerformed()`).

ssoloff commented on src/main/java/games/strategy/triplea/ui/TripleAFrame.java:
```diff
@@ -1979,6 +1979,44 @@ public void actionPerformed(final ActionEvent e) {
+   *        - list of producer territories to choose from
+   * @param unitTerritory
+   *        - territory units are being placed
+   * @return territory selected to use for production
```
Might want to document that `null` can be returned if the user does not make a selection.

Alternatively, you could return `Optional<Territory>`.

ssoloff commented on src/main/java/games/strategy/triplea/player/ITripleAPlayer.java:
```diff
@@ -166,6 +166,17 @@ Unit whatShouldBomberBomb(Territory territory, final Collection<Unit> potentialT
+   *        - a collection of Territories, the possible territories to produce units
+   * @param unitTerritory
+   *        - where units are being placed
+   * @return the territory selected as the producer
```
Might want to document that `null` can be returned if no territory is selected.

Alternatively, you could return `Optional<Territory>`.

ssoloff commented on src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:
```diff
@@ -204,23 +204,36 @@ public String placeUnits(final Collection<Unit> units, final Territory at, final
+      if (producers.size() > 1) {
+        producer = getRemotePlayer().selectProducerTerritoryForUnits(producers, at);
+        if (producer == null) {
+          break;
```
When I cancel the "Select Producer Territory" dialog, the following message is displayed:

![pr-1766-cancel](https://cloud.githubusercontent.com/assets/4826349/26558416/83fb7f08-4476-11e7-8de8-3df48f2f8c19.png)

Not sure if this is the intended behavior.  It doesn't seem to have any harmful side effects, though.

ron-murhammer commented:
@ssoloff Updated based on your comments. The message dialog is intended as the path for canceling or unit placement failing both lead there. I updated the text to make more sense.

Merging this now.
